pandas/core/frame.py

```diff
@@ -6892,56 +6892,97 @@ def cov(self, min_periods=None):
+            * kendall : Kendall Tau correlation coefficient
+            * spearman : Spearman rank correlation
+            * callable : callable with input two 1d ndarrays
+                 and returning a float
+            .. versionadded:: 0.24.0
+        min_periods : int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result. Currently only available for pearson
+            and spearman correlation
+            .. versionadded:: 0.24.0
+        Returns
+        -------
+        y : DataFrame
 
         Returns
         -------
-        correls : Series
+        correl : Series (how='pairwise') or DataFrame (how='all')
         """
+
         axis = self._get_axis_number(axis)
         this = self._get_numeric_data()
 
         if isinstance(other, Series):
-            return this.apply(other.corr, axis=axis)
+            return this.apply(other.corr, axis=axis,
+                              method=method, min_periods=min_periods)
 
-        other = other._get_numeric_data()
+        if not isinstance(other, DataFrame):
+            raise TypeError("'other' parameter should be a DataFrame "
+                            "or a Series object")
 
-        left, right = this.align(other, join='inner', copy=False)
+        other = other._get_numeric_data()
 
-        # mask missing values
-        left = left + right * 0
-        right = right + left * 0
+        if isinstance(this.columns, MultiIndex) \
+                or isinstance(other.columns, MultiIndex):
+            raise TypeError("MultiIndex is not supported")
 
         if axis == 1:
-            left = left.T
-            right = right.T
+            this = this.transpose()
+            other = other.transpose()
+
+        if how == 'all':
+            corr = np.zeros((this.shape[1], other.shape[1]))
+
+            for i, col1 in enumerate(this.columns):
+                for j, col2 in enumerate(other.columns):
+                    corr[i, j] = this.loc[:, col1].corr(
+                        other.loc[:, col2],
+                        method=method, min_periods=min_periods)
+            return DataFrame(data=corr, index=this.columns,
+                             columns=other.columns)
+
+        elif how == 'pairwise':
+            index = []
+            corr = []
+            for col in this.columns:
+                if col in other.columns:
+                    index.append(col)
+                    corr.append(this.loc[:, col].corr(other.loc[:, col],
+                                method=method, min_periods=min_periods))
+
+            correl = Series(data=corr, index=index)
+
+            if not drop:  # add missing columns to the resulting Series
+                result_index = this._get_axis(1).union(other._get_axis(1))
+                correl = correl.reindex(result_index)
+            return correl
 
-        # demeaned data
-        ldem = left - left.mean()
-        rdem = right - right.mean()
-
```

Comment:
benchmark you new impl vs the case that hits this. I can't imagine that iterating over the columns is better than this with any real number of columns. eg.. try 1000

Reply:
>this looks extremely fragile. pls align before you do anything like interating. then intersect the indexes first.

What to you want to align? As I said indexes are aligned automatically in `pd.Series.corr()`:
```python
def corr(self, other, method='pearson', min_periods=None):
    this, other = self.align(other, join='inner', copy=False)
    if len(this) == 0:
        return np.nan

    if method in ['pearson', 'spearman', 'kendall']:
        return nanops.nancorr(this.values, other.values, method=method,
                              min_periods=min_periods)
```

Reply:
30 columns:
https://github.com/pandas-dev/pandas/pull/22622#issuecomment-427882357

3000 columns:
https://github.com/pandas-dev/pandas/pull/22622#issuecomment-427893191